main/driver_setup.c

```diff
@@ -0,0 +1,409 @@
+static void can_recovery_task(void *pvParameters);
+
+// Prints the status of `netif` to `buf` in JSON format.
+// Returns the number of bytes written, or -1 if ran out of space.
+static int print_netif_status(esp_netif_t *netif, char *buf_out, size_t buflen);
+
+void driver_setup_ethernet(const esp_netif_ip_info_t *ip_info) {
+  // Based on:
+  // https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/network/esp_eth.html
+
+  // Check if the ethernet driver has already been started
+  if (driver_setup_eth_netif != NULL) {
+    ESP_LOGE(TAG, "Can only call driver_setup_ethernet() one time. Aborting.");
+    abort();
+  }
+
+  //// Create an ethernet MAC object ////
+  eth_mac_config_t mac_config = ETH_MAC_DEFAULT_CONFIG();
+  eth_esp32_emac_config_t esp32_emac_config = ETH_ESP32_EMAC_DEFAULT_CONFIG();
+  // 23 nased on ESP32-EVB schematic
+  esp32_emac_config.smi_mdc_gpio_num = 23;
+  // 18 based on ESP32-EVB schematic
+  esp32_emac_config.smi_mdio_gpio_num = 18;
+  esp_eth_mac_t *mac = esp_eth_mac_new_esp32(&esp32_emac_config, &mac_config);
+  assert(mac);
+
+  //// Create an ethernet PHY object ////
+  eth_phy_config_t phy_config = ETH_PHY_DEFAULT_CONFIG();
+  // ESP32-EVB schematic says SMI address is 0x00
+  phy_config.phy_addr = 0;
+  // ESP32-EVB schematic says PHY reset is unconnected
+  phy_config.reset_gpio_num = -1;
+  esp_eth_phy_t *phy = esp_eth_phy_new_lan87xx(&phy_config);
+  assert(phy);
+
+  //// Get an eth_handle by installing the driver ////
+  esp_eth_config_t eth_config = ETH_DEFAULT_CONFIG(mac, phy);
+  esp_eth_handle_t eth_handle = NULL;
+  ESP_ERROR_CHECK(esp_eth_driver_install(&eth_config, &eth_handle));
+
+  //// Create netif object ////
+  esp_netif_config_t netif_config = ESP_NETIF_DEFAULT_ETH();
+  esp_netif_t *esp_netif = esp_netif_new(&netif_config);
+  assert(esp_netif);
+
+  //// Set static IP info if needed ////
+  if (ip_info != NULL) {
+    ESP_ERROR_CHECK(esp_netif_dhcpc_stop(esp_netif));
+    ESP_ERROR_CHECK(esp_netif_set_ip_info(esp_netif, ip_info));
+  }
+
+  // Register event handlers for debugging purposes
+  ESP_ERROR_CHECK(esp_event_handler_register(ETH_EVENT, ESP_EVENT_ANY_ID,
+                                             &ethernet_event_handler, NULL));
+  ESP_ERROR_CHECK(esp_event_handler_register(IP_EVENT, IP_EVENT_ETH_GOT_IP,
+                                             &ip_event_handler, NULL));
+
+  //// Attach the ethernet object to ESP netif ////
+  esp_eth_netif_glue_handle_t eth_netif_glue =
+      esp_eth_new_netif_glue(eth_handle);
+  ESP_ERROR_CHECK(esp_netif_attach(esp_netif, eth_netif_glue));
+
+  //// Start ethernet ////
+  internet_ready = xSemaphoreCreateBinary();
+  ESP_ERROR_CHECK(esp_eth_start(eth_handle));
+
+  /// Wait for internet to actually work ////
+  if (xSemaphoreTake(internet_ready, pdMS_TO_TICKS(20000)) == pdFALSE) {
+    ESP_LOGE(TAG,
+             "Couldn't connect to the internet for 20 seconds. Restarting.");
+    abort();
```

Comment:
`abort()` seems a bit heavy handed to me to try to connect to Ethernet again. Consider just trying to set up Ethernet again?

Reply:
Fixed this by making the function recoverable in 041181b.